docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,564 @@
+- If an error occurs, `ControlPlane.Status.ErrorStatus` and `ControlPlane.Status.ErrorMessage` are populated.
+- `ControlPlane.Spec.Replicas` must be one of {1,3,5,7}.
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
+- The kubeadm bootstrapping configuration provided via `ControlPlane.Spec.KubeadmConfigSpec` should specify the `InitConfiguration`, `ClusterConfiguration`, and `JoinConfiguration` stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate.
+- The underlying query used to find existing Control Plane Machines is based on the following hardcoded label selector:
+
+```yaml
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      cluster.x-k8s.io/controlplane: "true"
+```
+
+- Generate CA certificates if they do not exist
+- Generate the kubeconfig secret if it does not exist
+
+Given the following `cluster.yaml` file:
+
+```yaml
+kind: Cluster
+apiVersion: cluster.x-k8s.io/v1alpha3
+metadata:
+  name: my-cluster
+  namespace: default
+spec:
+    clusterNetwork:
+      pods:
+        cidrBlocks: ["192.168.0.0/16"]
+  controlPlaneRef:
+    kind: ControlPlane
+    apiVersion: cluster.x-k8s.io/v1alpha3
+    name: my-controlplane
+    namespace: default
+  infrastructureRef:
+    kind: AWSCluster
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+    name: my-awscluster
+    namespace: default
+---
+kind: ControlPlane
+apiVersion: cluster.x-k8s.io/v1alpha3
+metadata:
+  name: my-control-plane
+  namespace: default
+spec:
+  replicas: 1
+  version: v1.16.0
+  infrastructureTemplate:
+    kind: AWSMachineTemplate
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+    namespace: default
+    name: my-awsmachinetemplate
+  kubeadmConfigSpec:
+    initConfiguration:
+      nodeRegistration:
+        name: '{{ ds.meta_data.hostname }}'
+        kubeletExtraArgs:
+          cloud-provider: aws
+    clusterConfiguration:
+      apiServer:
+        extraArgs:
+          cloud-provider: aws
+      controllerManager:
+        extraArgs:
+          cloud-provider: aws
+    joinConfiguration:
+      controlPlane: {}
+      nodeRegistration:
+        name: '{{ ds.meta_data.hostname }}'
+        kubeletExtraArgs:
+          cloud-provider: aws
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: AWSMachineTemplate
+metadata:
+  name: my-awsmachinetemplate
+  namespace: default
+spec:
+  ami:
+    id: ami-123456abcdef
+  instanceType: t3.large
+  iamInstanceProfile: "control-plane.cluster-api-provider-aws.sigs.k8s.io"
+  sshKeyName: my-ssh-key
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: AWSCluster
```

Comment:
Would it make sense to make this more general? Maybe we can use a "fake" provider name like `SomeProviderCluster` since this is in the cluster-api repo we should avoid specific  infra-provider references in proposals IMO (unless we're giving examples for the other providers too). I know it's meant to be an example for illustration purposes but using a specific reference doesn't encourage the reader to have other infra providers in mind when reading the proposal.

Reply:
Good point. The concrete schema is helpful for discussion, but we could easily replace some terms (e.g. AWS, ami, region), but keep the schema.

Reply:
This has now been updated, thanks for the feedback.